### PR TITLE
C regression tests

### PR DESCRIPTION
**Pull request by lsk567**
This PR makes the following changes:
1. Add a new directory programs to host all the test programs.
2. Update simify_verilog.py to load instruction files (.hex) on the command line.
3. Add a few C tests, including add, malloc, and calloc.
4. Make DSPM larger and simulation cycle longer to be able to handle larger C programs.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "riscv-tests"]
+	path = programs/tests/riscv
+	url = git@github.com:riscv-software-src/riscv-tests.git
```
Comment (edwardcwang) [inline .gitmodules:3]: If this is a public repo it should use the https link (otherwise when cloning on a keyless machine it may give errors)

File: src/test/scala/core/RegisterFileTest.scala [hunk 2]
```diff
@@ -25,10 +25,10 @@ class RegisterFileTest extends FlatSpec with ChiselScalatestTester {
   def registerFile = new RegisterFile(threads=threads)
 
   /* Write something to the register file */
-  def write(c: Module, w: RegisterFileWriteIO, thread: Int, addr: Int, data: UInt): Unit = {
+  def write(c: Module, w: RegisterFileWriteIO, enable: Bool, thread: Int, addr: Int, data: UInt): Unit = {
     require(addr < 32)
     timescope {
-      w.enable.poke(true.B)
+      w.enable.poke(enable)
```
Comment (edwardcwang) [inline src/test/scala/core/RegisterFileTest.scala:31]: In addition to this we should probably also poke the default for enable to be `false.B`

File: scripts/c/riscv_compile.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+# - Shaokai Lin <shaokai@eecs.berkeley.edu>
+
+set -ex
+set -euo pipefail
+
+SCRIPT_DIR=$(dirname "$0")
+LIB_DIR=$SCRIPT_DIR/../../programs/lib
+
+BIN="$1"
+shift
```
Comment (edwardcwang) [inline scripts/c/riscv_compile.sh:14]: ?

File: scripts/hdl/simify_verilog.py [hunk 4]
```diff
@@ -21,11 +21,17 @@ def add_ispm(contents: str, filename: str) -> str:
     assert m is not None
 
     orig_str = m.group()
-    new_str = f"{orig_str} initial $readmemh(\"{filename}\", ispm);"
+    new_str = f"""
+initial begin
+  $value$plusargs("ispm=%s", hex_file_name);
+  $readmemh(hex_file_name, ispm);
```
Comment (edwardcwang) [inline scripts/hdl/simify_verilog.py:27]: ?

File: src/main/scala/Core/RegisterFile.scala [hunk 5]
```diff
@@ -66,14 +66,16 @@ class RegisterFile(val threads: Int, val readPorts: Int = 2, val writePorts: Int
     // We also need to register the data and addresses since we are returning
     // last cycle's requests.
     val readUnderWrites = (writeIndexes zip io.write).map { case (writeIndex, writePort) =>
-      (RegNext(writeIndex) -> RegNext(writePort.data))
+      (RegNext(writeIndex) ->
```
Comment (edwardcwang) [inline src/main/scala/Core/RegisterFile.scala:69]: This looks like an unrelated change

Comment (lsk567) [reply src/main/scala/Core/RegisterFile.scala:69]: Right... These are the changes from #31, which the current branch is built on top of. Do you think we should combine these PRs into one PR, or keep them separate?

File: scripts/hdl/simify_verilog.py (hunk 4 above)
Comment (lsk567) [reply scripts/hdl/simify_verilog.py:27]: The idea here is that we can now pass the hex file into the emulator using command line arguments. So that the test harness can reuse the same emulator to run different programs with distinct filenames (as opposed to `imem.hex.txt` only). Is this a good way to implement this feature?

Comment (edwardcwang) [reply scripts/hdl/simify_verilog.py:27]: Isn't that what was already happening?

File: src/test/scala/core/RegisterFileTest.scala (hunk 2 above)
Comment (lsk567) [reply src/test/scala/core/RegisterFileTest.scala:31]: Sounds good. Will add that to the test.

File: scripts/hdl/simify_verilog.py (hunk 4 above)
Comment (lsk567) [reply scripts/hdl/simify_verilog.py:27]: On the master branch, the emulator only recognizes a hex file with the name `imem.hex.txt`. The change above allows it to run different programs by passing in their filenames without having to rename any programs as `imem.hex.txt`.

So if we have three test programs, `a.hex`, `b.hex`, and `c.hex`. To run them, we just need to do
```
flexpret-emulator +ispm=a.hex
flexpret-emulator +ispm=b.hex
flexpret-emulator +ispm=c.hex
```
without having to change any of the hex files to `imem.hex.txt` first.

File: azure-pipelines.yml [hunk 6]
```diff
@@ -28,5 +28,5 @@ steps:
   displayName: 'Unit tests'
 - script: ./test/basic_build.sh
   displayName: 'Basic build'
-- script: (export PATH="$HOME/riscv-embed-gcc/bin:$PATH" && cd ./programs/tests/ && make run-c-tests
+- script: (export PATH="$HOME/riscv-embed-gcc/bin:$PATH" && cd ./programs/tests/ && make run-c-tests)
```
Comment (lsk567) [inline azure-pipelines.yml:31]: Not sure why the CI test is failing...
```
2022-02-17T19:43:30.0497312Z ##[section]Starting: Run C tests
2022-02-17T19:43:30.0508002Z ==============================================================================
2022-02-17T19:43:30.0508261Z Task         : Command line
2022-02-17T19:43:30.0508543Z Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
2022-02-17T19:43:30.0508801Z Version      : 2.198.0
2022-02-17T19:43:30.0508972Z Author       : Microsoft Corporation
2022-02-17T19:43:30.0509252Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
2022-02-17T19:43:30.0509570Z ==============================================================================
2022-02-17T19:43:30.1662826Z Generating script.
2022-02-17T19:43:30.1683601Z Script contents:
2022-02-17T19:43:30.1684695Z (export PATH="$HOME/riscv-embed-gcc/bin:$PATH" && cd ./programs/tests/ && make run-c-tests)
2022-02-17T19:43:30.1685317Z ========================== Starting Command Output ===========================
2022-02-17T19:43:30.1692907Z [command]/usr/bin/bash --noprofile --norc /home/vsts/work/_temp/56650c1f-22e4-4a3f-b9ac-2599f134ce6d.sh
2022-02-17T19:43:30.1752841Z $base_dir value -- /home/vsts/work/1/s
2022-02-17T19:43:30.1753769Z $generated_dir value -- /home/vsts/work/1/s/out
2022-02-17T19:43:30.1754781Z $emu_dir value -- /home/vsts/work/1/s/emulator
2022-02-17T19:43:30.1755532Z $tests_dir value -- /home/vsts/work/1/s/programs/tests
2022-02-17T19:43:30.1763270Z cp /home/vsts/work/1/s/programs/tests/c-tests/add /home/vsts/work/1/s/programs/tests/c-tests/add.out
2022-02-17T19:43:30.1779292Z cp: -r not specified; omitting directory '/home/vsts/work/1/s/programs/tests/c-tests/add'
2022-02-17T19:43:30.1782629Z make: *** [<builtin>: /home/vsts/work/1/s/programs/tests/c-tests/add.out] Error 1
2022-02-17T19:43:30.1833605Z ##[error]Bash exited with code '2'.
2022-02-17T19:43:30.1866422Z ##[section]Finishing: Run C tests
```
On my local machine, there is no `cp` operation.

Comment (lhstrh) [reply azure-pipelines.yml:31]: `cp /home/vsts/work/1/s/programs/tests/c-tests/add /home/vsts/work/1/s/programs/tests/c-tests/add.out` won't work because the first argument is a directory, not a file. Either you have to specify a `-r` or `-a` flag, or fix the path so that it points to a file.

Comment (lsk567) [reply azure-pipelines.yml:31]: Problem resolved. Looks like the pattern rule was incorrect.
